src/InvoiceManager.php

```diff
@@ -0,0 +1,326 @@
+<?php
+
+namespace dnj\Invoice;
+
+use dnj\Invoice\Contracts\IInvoice;
+use dnj\Invoice\Contracts\IInvoiceManager;
+use dnj\Invoice\Contracts\InvoiceStatus;
+use dnj\Invoice\Contracts\IPayment;
+use dnj\Invoice\Contracts\IProduct;
+use dnj\Invoice\Contracts\PaymentStatus;
+use dnj\Invoice\Exceptions\AmountInvoiceMismatchException;
+use dnj\Invoice\Exceptions\CurrencyMismatchException;
+use dnj\Invoice\Exceptions\InvalidInvoiceStatusException;
+use dnj\Invoice\Exceptions\InvoiceUserMismatchException;
+use dnj\Invoice\Models\Invoice;
+use dnj\Invoice\Models\Payment;
+use dnj\Invoice\Models\Product;
+use dnj\Invoice\traits\ProductBuilder;
+use dnj\Number\Contracts\INumber;
+use dnj\Number\Number;
+use Illuminate\Support\Facades\DB;
+
+class InvoiceManager implements IInvoiceManager
+{
+    use ProductBuilder;
+
+    public function create(int $userId, int $currencyId, array $products, array $localizedDetails, ?array $meta): IInvoice
+    {
+        // TODO: Implement create() method.
+
+        $invoice = new Invoice();
+        $invoice->user_id = $userId;
+        $invoice->currency_id = $currencyId;
+        $invoice->meta = $meta;
+        if (isset($localizedDetails['title'])) {
+            $invoice->title = $localizedDetails['title'];
+        }
+        $invoice->status = InvoiceStatus::UNPAID->value;
+        $invoice->save();
+        $invoice->products()
+                ->createMany($this->buliderInvoiceProducts($products));
+        $invoice->update([
+                             'amount' => Number::fromInt($invoice->products->sum('total_amount'))
+                                               ->getValue(),
+                         ]);
+
+        return $invoice;
+    }
+
+    public function delete(int $invoiceId): void
+    {
+        // TODO: Implement delete() method.
+
+        $invoice = $this->getInvoiceById($invoiceId);
+        if (InvoiceStatus::PAID == $invoice->status) {
+            throw new InvalidInvoiceStatusException();
+        }
+        if ($invoice->products->isNotEmpty()) {
+            $invoice->products()
+                    ->delete();
+        }
+        if ($invoice->payments->isNotEmpty()) {
+            $invoice->payments()
+                    ->delete();
+        }
+        $invoice->delete();
```

Comment:
just make relations cascade on delete and delete invoice. no need to delete payments and products separately.

Reply:
done